src/main/kotlin/uk/gov/dluhc/printapi/service/PrintResponseProcessingService.kt

```diff
@@ -0,0 +1,92 @@
+    private val idFactory: IdFactory,
+    private val clock: Clock
+) {
+    /**
+     * Updates print details for the batches based on the batch responses' statuses.
+     * If a batch is successful, a new printRequestStatus will be added for each corresponding print details with status
+     * being RECEIVED_BY_PRINT_PROVIDER.
+     *
+     * If a batch is not successful, the status will be set to SENT_TO_PRINT_PROVIDER and a new requestId will be set.
+     * The reason for the new requestId is, the print provider does not support duplicate print requests with the same requestId.
+     *
+     * dateCreated will be the timestamp the new printRequestStatus is created and eventDateTime will be the timestamp from
+     * the print provider's batch response.
+     */
+    fun processBatchResponses(batchResponses: List<BatchResponse>) {
+        batchResponses.forEach { batchResponse ->
+            with(batchResponse) {
+                val newStatus =
+                    if (status == SUCCESS) RECEIVED_BY_PRINT_PROVIDER else PENDING_ASSIGNMENT_TO_BATCH
+
+                val printDetails =
+                    printDetailsRepository.getAllByStatusAndBatchId(SENT_TO_PRINT_PROVIDER, batchId)
+
+                printDetails.forEach {
+                    it.addStatus(
+                        status = newStatus,
+                        dateCreated = OffsetDateTime.now(clock),
+                        eventDateTime = timestamp,
+                        message = null
+                    )
+
+                    if (status != SUCCESS) {
+                        it.requestId = idFactory.requestId()
+                    }
+                }
+
+                printDetailsRepository.updateItems(printDetails)
+            }
+        }
+    }
+
+    fun processPrintResponse(printResponse: ProcessPrintResponseMessage) {
+        val newStatus = with(printResponse) {
+            when {
+                statusStep == ResponseStatusStep.PROCESSED && status == ResponseStatus.SUCCESS -> VALIDATED_BY_PRINT_PROVIDER
+                statusStep == ResponseStatusStep.IN_MINUS_PRODUCTION && status == ResponseStatus.SUCCESS -> IN_PRODUCTION
+                statusStep == ResponseStatusStep.DISPATCHED && status == ResponseStatus.SUCCESS -> DISPATCHED
+                statusStep == ResponseStatusStep.NOT_MINUS_DELIVERED && status == ResponseStatus.FAILED -> NOT_DELIVERED
+                statusStep == ResponseStatusStep.PROCESSED && status == ResponseStatus.FAILED -> PRINT_PROVIDER_VALIDATION_FAILED
+                statusStep == ResponseStatusStep.IN_MINUS_PRODUCTION && status == ResponseStatus.FAILED -> PRINT_PROVIDER_PRODUCTION_FAILED
+                statusStep == ResponseStatusStep.DISPATCHED && status == ResponseStatus.FAILED -> PRINT_PROVIDER_DISPATCH_FAILED
+                else -> throw IllegalArgumentException("Undefined statusStep [$statusStep] and status [$status] combination")
+            }
+        }
+
```

Comment:
These conditions are quite confusing. I have another plan/implementation. 
My version would be this to evaluate `newStatus`. Feel free to refactor this please.

```
val newStatus = with(printResponse) {
            when (status) {
                ResponseStatus.SUCCESS -> {
                    when (statusStep) {
                        ResponseStatusStep.PROCESSED -> VALIDATED_BY_PRINT_PROVIDER
                        ResponseStatusStep.IN_MINUS_PRODUCTION -> IN_PRODUCTION
                        ResponseStatusStep.DISPATCHED -> DISPATCHED
                        else -> throw IllegalArgumentException("Undefined statusStep [$statusStep] and status [$status] combination")
                    }
                }
                ResponseStatus.FAILED -> {
                    when (statusStep) {
                        ResponseStatusStep.NOT_MINUS_DELIVERED -> NOT_DELIVERED
                        ResponseStatusStep.PROCESSED -> PRINT_PROVIDER_VALIDATION_FAILED
                        ResponseStatusStep.IN_MINUS_PRODUCTION -> PRINT_PROVIDER_PRODUCTION_FAILED
                        ResponseStatusStep.DISPATCHED -> PRINT_PROVIDER_DISPATCH_FAILED
                    }
                }
            }
        }
```

Reply:
Ah jinx 😁 I was just writing the exact same suggestion and the exact same code snippet 😁 Great minds think alike 😁 👍 

The other idea/approach I was thinking was a MapStruct mapper? Would that be a better/cleaner approach? And it has the added bonus that we can write a unit test for the mapping 👍 

What do you think? 🤔

Reply:
Well tbh i also thought of a mapper here.. 
so if that's easier to mapstruct it then its even better 😊 (btw i m still not there, still catching up with your great mind @NathanRussellValtech)

Reply:
Thanks, @vishalgupta-valtech  and @NathanRussellValtech for your suggestions.
I will introduce a mapper accordingly.

Reply:
@NathanRussellValtech , I tried really hard to use `Mapstruct` for the mapping, but unfortunately couldn't find a way to map to an enum from two source values. Hence, the mapper is just using the suggested when implementation.